rllib/agents/slateq/slateq_tf_policy.py

```diff
@@ -0,0 +1,356 @@
+    stats = {
+        "q_values": policy._q_values,
+        "q_clicked": policy._q_clicked,
+        "scores": policy._scores,
+        "score_no_click": policy._score_no_click,
+        "slate_q_values": policy._slate_q_values,
+        "replay_click_q": policy._replay_click_q,
+        "bellman_reward": policy._bellman_reward,
+        "target": policy._target,
+        "next_q_target_slate": policy._next_q_target_slate,
+        "next_q_target_max": policy._next_q_target_max,
+        "target_clicked": policy._target_clicked,
+        "q_loss": policy._q_loss,
+    }
+    return stats
+
+
+def action_distribution_fn(
+    policy: Policy, model: SlateQTFModel, input_dict, *, explore, is_training, **kwargs
+):
+    """Determine which action to take."""
+
+    # First, we transform the observation into its unflattened form.
+    observation = restore_original_dimensions(
+        input_dict[SampleBatch.CUR_OBS], policy.observation_space, tensorlib=tf
+    )
+    # user.shape: [B, E]
+    user_obs = observation["user"]
+    doc_obs = list(observation["doc"].values())
+    slate_size = len(policy.action_space.nvec)
+
+    # Compute scores per candidate.
+    scores, score_no_click = score_documents(user_obs, doc_obs)
+    # Compute Q-values per candidate.
+    q_values = model.get_q_values(user_obs, doc_obs)
+
+    with tf.name_scope("select_slate"):
+        per_slate_q_values = get_per_slate_q_values(
+            policy.slates, slate_size, score_no_click, scores, q_values
+        )
+    model.slates = policy.slates
+    return per_slate_q_values, Categorical, []
+
+
+def get_per_slate_q_values(slates, slate_size, s_no_click, s, q):
+    slate_q_values = tf.gather(s * q, slates, axis=1)
+    slate_scores = tf.gather(s, slates, axis=1)
+    slate_normalizer = tf.reduce_sum(
+        input_tensor=slate_scores, axis=2
+    ) + tf.expand_dims(s_no_click, 1)
+
+    slate_q_values = slate_q_values / tf.expand_dims(slate_normalizer, 2)
+    slate_sum_q_values = tf.reduce_sum(input_tensor=slate_q_values, axis=2)
+    return slate_sum_q_values
+
+
+def score_documents(
+    user_obs, doc_obs, no_click_score=1.0, multinomial_logits=False, min_normalizer=-1.0
+):
+    """Computes dot-product scores for user vs doc (plus no-click) feature vectors."""
+
+    # Dot product between used and each document feature vector.
+    scores_per_candidate = tf.reduce_sum(
+        tf.multiply(tf.expand_dims(user_obs, 1), tf.stack(doc_obs, axis=1)), 2
+    )
+    # Compile a constant no-click score tensor.
+    score_no_click = tf.fill([tf.shape(user_obs)[0], 1], no_click_score)
+    # Concatenate click and no-click scores.
+    all_scores = tf.concat([scores_per_candidate, score_no_click], axis=1)
+
+    # Logits: Softmax to yield probabilities.
+    if multinomial_logits:
+        all_scores = tf.nn.softmax(all_scores)
+    # Multinomial proportional model: Shift to `[0.0,..[`.
+    else:
+        all_scores = all_scores - min_normalizer
```

Comment:
hmmm, this is also kind of very recsim specific.
I think at the very least, we need to document multinomial_logits=False and min_normalizer=-1.0 above well.